### ISSUE TITLE
Performance regression 2.28

### ISSUE THREAD

**Issue by rok-cesnovar** [stan-dev/cmdstan#1055]
#### Description:

A performance regression was identified on Discourse: https://discourse.mc-stan.org/t/speed-difference-between-rstan-and-c
cmdstan-for-a-simple-model/25113

The regression is most evident with a trivial model like: 
```stan
data {
 int N;
}
parameters {
 vector[N] a;
}
model {
 a ~ normal(0,1);
}
```

Execution times with N=10000:

![image](https://user-images.githubusercontent.com/28476796/140659466-3d0efe1a-7384-4523-b2e5-cf9c8862c55c.png)

It's much less evident with non-trivial models, but still, something to investigate.

Based on my investigation and investigation by the user `rfc` on Discourse (EDIT: @fcostin on Github)  it has to do with writing output samples: https://discourse.mc-stan.org/t/speed-difference-between-rstan-and-cmdstan-for-a-simple-model/25113/23

It's probably the changes made to work with multiple chains.

EDIT: https://github.com/stan-dev/cmdstan/pull/987 was confirmed to be the issue.
 

#### Current Version:
v2.28.1

**Comment by fcostin** [comment]
(i'm the same rfc from discourse)

I suspect it's the diagnostic output.  With cmdstan `2.27.0` if you explicitly enable diagnostic output on the command line, then it takes a similar amount of running time to cmdstan `2.28.1`.

Whereas the running time of cmdstan `2.28.1` doesn't change if the user specifies a diagnostic output.

Could it be possible that cmdstan `2.28.1` by default is doing all the hard work to format diagnostic output and then throw it away?

This part of #987 looks a bit suspicious: 
https://github.com/stan-dev/cmdstan/pull/987/files#diff-de5c2eb1977dd486b035c6a7ea3539d0b5c7bdac2cff55705b1344734a3a4f50R501-R503

**Comment by rok-cesnovar** [comment]
That is a very nice catch @fcostin!! You are spot on.

Commenting out all the diagnostic writes in the stan submodule gets the performance back. So we just need to figure out how to handle this in command.hpp.

**Comment by fcostin** [comment]
It looks like the culprit is  `stan/src/stan/callbacks/unique_stream_writer.hpp`:

https://github.com/stan-dev/stan/blob/0e4e2a9f8e4af6fcc6baae5070fc4acddbddd9ad/src/stan/callbacks/unique_stream_writer.hpp

In the case where the user does not specify a diagnostic_file, `command.hpp` is still setting the diagnostic writer to a `unique_stream_writer.hpp`.

Internally, `unique_stream_writer` is buffering all writes to a local `std::stringstream streamer` before writing to the output stream.

Curiously, removing the the `stringstream` buffering doesn't suffice to fix the problem.

If I add an explicit bool toggle to `unique_stream_writer` to set it into no-op mode, and modify `command.hpp` to toggle on no-op mode when constructing the diagnostic writer, then I can observe speedup and running time in line with cmdstan 2.27.0 performance.

**Comment by rok-cesnovar** [comment]
Thanks @fcostin for the suggested fix. 

I will let @SteveBronder chime in, as he know more about the decisions that went into this.

**Comment by wds15** [comment]
So is the performance regression due to buffering output after all? That's done to avoid cluttered output under threading. Hopefully we can make the buffer more efficient as this thing is there for a good reason.

Also: Is the performance regression seen for any model which does some more intense computations? Binomial likelihood with normalisation - how is that (e.g. the Bernoulli example model)? Or more complicated stuff like a negative binomial?

I am having the impression that almost trivial models are affected by this... which does not mean we should not fix it, but maybe not at "any cost"?

**Comment by rok-cesnovar** [comment]
> So is the performance regression due to buffering output after all?

Its due to buffering the output for diagnostics even when the diagnostics file is `""`. We essentially pay the cost of outputing diagnostics regardless if they are requested or not.

>  Is the performance regression seen for any model which does some more intense computations?

The more comutation there is, the less obvious the performance hit is.

> I am having the impression that almost trivial models are affected by this... which does not mean we should not fix it, but maybe not at "any cost"?

I agree that trivial models probably arent the focal point, but I think the solution here should be trivial. Worst case we add a boolean like @fcostin proposes.

The essential problem was that pre-2.28, this line produced a no-op diagnostic writer: https://github.com/stan-dev/cmdstan/blob/998c08fe8cea1b765f923ea30ce59e009459fc2f/src/cmdstan/command.hpp#L211
while this line https://github.com/stan-dev/cmdstan/blob/v2.28.1/src/cmdstan/command.hpp#L502 does not do that.

**Comment by wds15** [comment]
Sounds like an easy matter than to disable the diagnostic output whenever its not requested.

**Comment by fcostin** [comment]
i agree with @rok-cesnovar's summary.

my example patch is not minimal or the best solution -- there is no need to remove the buffering in `unique_stream_writer`. It would be cleaner to modify `command.hpp` so it disables diagnostic output by wiring in a no-op diagnostic writer. It should be easy but may need the type of the vector `diagnostic_writers` in `command.hpp` to be changed from a vector of `unique_stream_writer` to a vector of pointers to the base writer class

**Comment by wds15** [comment]
Oh dear…I reviewed this pr at the time and did not notice. Hopefully @SteveBronder can sort this out quickly.

**Comment by SteveBronder** [comment]
@fcostin think you for the thorough review! Yes I like your solution of making these a pointer up in `command.hpp` and having a `empty_writer` for making the writes a no-op.

**Comment by rok-cesnovar** [comment]
Closed by #1060
